env/env_encrypt2.cc

```diff
@@ -0,0 +1,572 @@
+// In case the implementation lists the directory prior to iterating the files
+// and files are concurrently deleted, the deleted files will be omitted from
+// result.
+// The name attributes are relative to "dir".
+// Original contents of *results are dropped.
+// Returns OK if "dir" exists and "*result" contains its children.
+//         NotFound if "dir" does not exist, the calling process does not have
+//                  permission to access "dir", or if "dir" is invalid.
+//         IOError if an IO Error was encountered
+Status EncryptedEnv2::GetChildrenFileAttributes(
+    const std::string& dir, std::vector<FileAttributes>* result) {
+  auto status = EnvWrapper::GetChildrenFileAttributes(dir, result);
+  if (status.ok()) {
+    // this is slightly expensive, but fortunately not used heavily
+    std::shared_ptr<EncryptionProvider> provider;
+
+    for (auto it = std::begin(*result); it != std::end(*result); ++it) {
+      status = GetEncryptionProvider(it->name, provider);
+
+      if (status.ok() && provider) {
+        size_t prefixLength = provider->GetPrefixLength();
+
+        if (prefixLength <= it->size_bytes) it->size_bytes -= prefixLength;
+      }
+    }
+  }
+
+  return status;
+}
+
+// Store the size of fname in *file_size.
+Status EncryptedEnv2::GetFileSize(const std::string& fname,
+                                  uint64_t* file_size) {
+  Status status;
+  status = EnvWrapper::GetFileSize(fname, file_size);
+
+  if (status.ok()) {
+    // this is slightly expensive, but fortunately not used heavily
+    std::shared_ptr<EncryptionProvider> provider;
+    status = GetEncryptionProvider(fname, provider);
+    if (status.ok() && provider) {
+      size_t prefixLength = provider->GetPrefixLength();
+      if (prefixLength <= *file_size) *file_size -= prefixLength;
+    }
+  }
+
+  return status;
+}
+
+Status EncryptedEnv2::GetEncryptionProvider(
+    const std::string& fname, std::shared_ptr<EncryptionProvider>& provider) {
+  std::unique_ptr<SequentialFile> underlying;
+  EnvOptions options;
+  Status status;
+
+  provider.reset();
+  status = Env::Default()->NewSequentialFile(fname, &underlying, options);
+
+  if (status.ok()) {
+    std::unique_ptr<BlockAccessCipherStream> stream;
+    status = EncryptedEnv2::ReadSeqEncryptionPrefix(underlying.get(), provider,
+                                                    stream);
+  }
+
+  return status;
+}
+
+UnixLibCrypto EncryptedEnv2::crypto_;
+
+Env* EncryptedEnv2::Default() {
+  // the rational for this routine is to help force the static
+  //  loading of UnixLibCrypto before other routines start
+  //  using the encryption code.
+  static EncryptedEnv2 default_env(Env::Default());
+  return &default_env;
+}
+
+Env* EncryptedEnv2::Default(EncryptedEnv2::ReadKeys encrypt_read,
+                            EncryptedEnv2::WriteKey encrypt_write) {
+  EncryptedEnv2* default_env = (EncryptedEnv2*)Default();
+  default_env->SetKeys(encrypt_read, encrypt_write);
```

Comment:
It looks like a bad pattern to modify internal data for a singleton.. But I understand `Options` only hold raw pointer of `Env`. Shall we add a `std::shared_ptr<Env> env_owned` field to `Options` so that the DB can manage the lifetime of a non-default `Env`?

@pdillinger what do you think from rocksdb team perspective?

Reply:
All access to encryption keys within the object are surrounded by a read/write lock.  It was going to be difficult to put the lock within the class due to RWMutex being hidden within port (i.e. not public).  I cheated and made it a local global.  It is not going to be hit often even if multiple databases are active.